File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/xpath/PseudoAttributeHelper.java
package com.github.jmlparser.xpath;

import com.github.javaparser.ast.Node;
import org.jetbrains.annotations.NotNull;
import org.w3c.dom.Element;

import java.util.Collection;

/**
 * @author Alexander Weigl
 * @version 1 (11.02.23)
 */
public abstract class PseudoAttributeHelper<T extends Node> implements PseudoAttributeProvider {
    private final Class<T> clazz;

    protected PseudoAttributeHelper(Class<T> clazz) {
        this.clazz = clazz;
    }

    @Override
    public final Collection<JPAttrPseudo> attributeForNode(@NotNull Node node, @NotNull Element owner) {
        if (clazz == node.getClass()) {
            return attributes((T) node, owner);
        }
        return null;
    }

    protected abstract Collection<JPAttrPseudo> attributes(@NotNull T node, Element owner);
}

File: tools/xpath/src/main/kotlin/io/github/jmltoolkit/xpath/PseudoAttributeHelper.kt
package io.github.jmltoolkit.xpath

import com.github.javaparser.ast.Node
import org.w3c.dom.Element

/**
 * @author Alexander Weigl
 * @version 1 (11.02.23)
 */
abstract class PseudoAttributeHelper<T : Node?> protected constructor(private val clazz: Class<T>) :
    PseudoAttributeProvider {
    @Suppress("UNCHECKED_CAST")
    override fun attributeForNode(node: Node, owner: Element): Collection<JPAttrPseudo> {
        if (clazz == node.javaClass) {
            return attributes(node as T, owner)
        }
        return listOf()
    }

    protected abstract fun attributes(node: T, owner: Element?): Collection<JPAttrPseudo>
}
